Emails, Codes, and Guides

firemaster1294 opened the pull request:
This PR does a LOT. I added the new player guide in as actual HTML. I added email visibility in the website so you don't need to actually check your email. And I added printable player codes :D however, this leaves a lot to be checked over before it can actually be pushed. Thanks for reviewing!

uwhvz committed:
Saves Emails for Player View

A longwinded way of saving data sent in emails for convenient access in the dashboard

uwhvz committed:
Aesthetics

Modified emails to allow players to see all sent emails when game is not running

uwhvz committed:
Enables saving of player made emails

uwhvz committed:
Implements the New Player Guide

uwhvz committed:
Creates Printable Player Codes

:D

uwhvz committed:
Compatibility fix

tig567899 commented on app/models/email.py:
```diff
@@ -0,0 +1,36 @@
+
+from .game import Game
+
+class EmailRule(Enum):
```
I really don't like the naming here. `EmailRule` implies that there's a rule about the email, not the the specified recipient group. I think something like `RecipientGroup` or `RecipientCode` would be better here.

firemaster1294 replied:
Renamed to RecipientGroup in most recent commit

tig567899 commented on app/templates/jinja2/dashboard/player.html:
```diff
@@ -11,7 +11,7 @@
 <div class="ui-grid s-dashboard__index">
   <div class="ui-grid">
-    {{ card("My Code", participant.code, "ui-card--display") }}
+    {{ card("My Code", <a href="{{ url('player_code) }}">participant.code</a>, "ui-card--display") }}
```
Just off a cursory glance, `'player_code'`. You're missing the ending single quote.

firemaster1294 replied:
Ty!

tig567899 commented on app/templates/jinja2/dashboard/player.html:
```diff
@@ -83,4 +83,4 @@
   {% endif %}
 </div>
-{% endblock %}
+{% endblock %}
```
Nit: I think it's probably best to keep the ending newlines in the files.

tig567899 commented on app/views/dashboard.py:
```diff
@@ -9,7 +9,7 @@
 
 from app.util import MobileSupportedView, most_recent_game
-from app.models import Game, Tag, Player, PlayerRole, SignupLocation, Legacy, SupplyCode, User
+from app.models import Game, Tag, Player, PlayerRole, SignupLocation, Legacy, SupplyCode, User, TagType, Email, EmailRule
```
You need to add all these things to the `__init__.py` files in the models folder. Otherwise they will not build.

tig567899 commented on app/models/email.py:
```diff
@@ -0,0 +1,36 @@
+class Email(models.Model):
+    id: uuid = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    game: Game = models.ForeignKey(Game, on_delete=models.PROTECT)
+    name: str = models.CharField()
+    data: str = models.CharField()
```
Running into `SystemCheckError` when trying to build this.
```
app.Email.data: (fields.E120) CharFields must define a 'max_length' attribute.
app.Email.name: (fields.E120) CharFields must define a 'max_length' attribute.
```

tig567899 commented on app/views/player.py:
```diff
@@ -9,13 +9,21 @@
 
+import io
+from django.http import FileResponse
+from reportlab.platypus import Table, TableStyle, SimpleDocTemplate
```
This introduces dependencies. I don't know exactly where we put our mandatory dependencies, but there is definitely a place for it. Perhaps @pkrshnan knows more about that.

firemaster1294 replied:
This is the only thing that I'm uncertain of thus far. I know we have the dependencies installed on the backend, but other than that I don't know what else we need to do

tig567899 reviewed (changes requested):
As of right now, this branch can't be built, so I can't test it. I recommend pulling from master so that you have the up-to-date changes, because this branch, even with rudimentary fixes for testing, is missing the critical fixes from #51.

I'll have a second look-over once my comments are addressed. Also, fix your merge conflicts.

uwhvz committed:
Merge remote-tracking branch 'origin/master' into Email-Saving

uwhvz committed:
Email formating corrections

tig567899 commented on app/views/player.py:
```diff
@@ -9,14 +9,22 @@
 
 from app.mail import send_tag_email, send_stun_email
-from app.models import Player, PlayerRole, Tag, SupplyCode, Modifier, ModifierType, Spectator, Moderator, TagType
+from app.models import Player, PlayerRole, Tag, SupplyCode, Modifier, ModifierType, Spectator, Moderator, TagType, Email, EmailRule
```
Build still fails here. You forgot to change EmailRule.

tig567899 commented on app/templates/jinja2/dashboard/player.html:
```diff
@@ -11,7 +11,7 @@
 <div class="ui-grid s-dashboard__index">
   <div class="ui-grid">
-    {{ card("My Code", participant.code, "ui-card--display") }}
+    {{ card("My Code", <a href="{{ url('player_code') }}">participant.code</a>, "ui-card--display") }}
```
This doesn't work. Gets error "Unexpected `<` character"

tig567899 replied:
From what I can tell, this is not designed to be linkable.

uwhvz committed:
Made corrections to email groupings

Also player code display

tig567899 commented on app/templates/jinja2/dashboard/player.html:
```diff
@@ -11,7 +11,9 @@
   <div class="ui-grid">
-    {{ card("My Code", participant.code, "ui-card--display") }}
+    {% call card("My Code") %}
+    <a href="{{ url('player_code') }}">participant.code</a>
```
![image](https://user-images.githubusercontent.com/17097380/90988332-044d0900-e560-11ea-9c46-7cc4f9f8b3a2.png)

Bruh.

Also, the link doesn't work.

firemaster1294 replied:
Oof my bad. There’s a relatively simply fix that I can push later today

uwhvz committed:
Fixed moderator usage

uwhvz committed:
Update dashboard.py

tig567899 commented on app/views/dashboard.py:
```diff
@@ -33,11 +33,20 @@ class DashboardView(MobileSupportedView):
+        
+        if participant.is_player and game.is_running:
+            if participant.is_human:
+                emails = Email.objects.filter(game=game).exclude(rule=RecipientGroup.ZOMBIE)
+            elif participant.is_zombie:
+                emails = Email.objects.filter(game=game).exclude(rule=RecipientGroup.HUMAN)
```
Both of these should be changed to `group=`

tig567899 commented on app/views/dashboard.py:
```diff
@@ -33,11 +33,20 @@ class DashboardView(MobileSupportedView):
 
     def get(self, request):
         game = most_recent_game()
+        participant = request.user.participant(game)
```
For some reason, this line returns `NoneType` for superuser accounts. We should investigate this later, but for now, it's causing line 43 to crash when a superuser logs in.

tig567899 commented on app/templates/jinja2/dashboard/index.html:
```diff
@@ -23,6 +23,90 @@
+          {% for email in emails.exclude(player_made) %}
+            <tr>
+              <td data-toggle="tooltip" data-placement="top" title="{{ email.name }}">{{ email.name }}</td>
+              <td>{{ email.created_at }}</td>
```
This line is erroring. I'm not sure by what, but the error received is the following:
![image](https://user-images.githubusercontent.com/17097380/91878467-56bbb300-ec4d-11ea-9187-8db1c2cdc100.png)

There is probably an error in here somewhere.

tig567899 commented on app/views/player.py:
```diff
@@ -33,6 +41,40 @@ def render_player_info(request, report_tag_form=ReportTagForm(), claim_supply_co
+    
+        # Create the PDF object, using the buffer as its "file."
+        p = SimpleDocTemplate(buffer, topMargin=1)
+        width, height = letter # Save these for later
```
`letter` is undefined. Causes this view to crash.

uwhvz committed:
Email tweaks

uwhvz committed:
Added Minecraft

uwhvz committed:
Fixed letter error

uwhvz committed:
UI Changes and Improvements

Changed dependencies to include reportlab

Added public score functionality

Corrected total stun/kill counter

General UI Changes

tig567899 reviewed (changes requested):
Generally I like the new changes, but there are a few things that need to be fixed before I approve.

- It's not obvious that the participant code is clickable. As a player, it's very likely that I miss it. As well, I don't like the fact that it automatically downloads. If it can open a PDF in my browser, that'd be much better.
- On the new player page, the dropdowns look like navigable links. I was confused for a while before I figured out that I needed to click on them. I would add a down arrow to them to provide an affordance on that link.
- The fonts are different on the new player page for the different sections.
- On a related nit: The "Recommended Against" subtitle is not obvious that it's a subtitle.

uwhvz committed:
Improved Player Code Downloads

uwhvz committed:
Added Drop Down Bar for Panels

uwhvz committed:
Cosmetic changes

Fixed drop down menus and font sizes

tig567899 commented:
Looks good but a few minor issues.

1. The bottom spacing for signups is larger than everything else. As well, when expanded, there's a white line separating the title and text. No other heading has this behaviour.
2. Expanding/minimizing subheadings also rotates the arrow of the parent heading. This is unintuitive. If I were to guess, this is likely because the two elements share a class or similar.

uwhvz committed:
Even more cosmetics

uwhvz committed:
Merge branch 'master' into Email-Saving

uwhvz committed:
Update navbar.html
